sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileBasedSink.java

```diff
@@ -229,7 +229,7 @@ public static ResourceId convertToFileResourceIfPossible(String outputPrefix) {
       implements HasDisplayData, Serializable {
     /**
      * Returns an object that represents at a high level the destination being written to. May not
-     * return null.
+     * return null. A destination must have deterministic hash and equality methods defined.
```

Comment:
Same reason we required it in BigQueryIO's version - we often create hashmaps maps keyed by DestinationT.  Yes  - we could always serialize the destination and hash bytes each time, however requiring hashCode() seems like a reasonable thing, and saves expense and code complexity.